### PR TITLE
Register vpc, subnet ID and task-eni attributes

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

Registers container instance with vpc and subnet ID attributes.

### Implementation details

* de1fbad agent/ec2: extended the metadata interface to retrieve subnet and vpc id's
* 19b8132 agent/api/ecsclient: register instance with vpc and subnet ID attributes
* ec1bf6a agent/app, agent/engine: modify methods that use DockerTaskEngine's Capability method as per changes in the method signature
* 2812f0d agent/api, agent/engine: get rid of GetAdditionalAttributes method in TaskEngine
* 59787c4 agent/app,agent/api: reorder eni watcher initialization in agent.go
* c938965 agent/engine: register the 'task-eni' capability only if all plugins support it
* 9a23726 agent/ecscni: modify the CNIPlugin interface to retrieve capabilites of a plugin

### Testing

- [X] Builds on Linux (`make release`)
- [X] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [X] Unit tests on Linux (`make test`) pass
- [X] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [X] Integration tests on Linux (`make run-integ-tests`) pass
- [X] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [X] Functional tests on Linux (`make run-functional-tests`) pass
- [X] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: Yes

### Description for the changelog

enhancement - register the Agent with vpc and subnet ID's

### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

File: agent/engine/docker_task_engine.go [hunk 1]
```diff
@@ -936,44 +945,34 @@ func (engine *DockerTaskEngine) isParallelPullCompatible() bool {
...
-		for _, plugin := range plugins {
-			_, err := engine.cniClient.Version(plugin)
-			if err != nil {
-				log.Error("Engine: Check version of plugin %s failed", plugin)
-				return nil, false
-			}
+	for _, plugin := range plugins {
+		// Check if we can get version information from each plugin
+		version, err := engine.cniClient.Version(plugin)
+		if err != nil {
```
Comment (richardpen) [inline agent/engine/docker_task_engine.go]: Based on the implementation [here](https://github.com/aws/amazon-ecs-cni-plugins/blob/dev/plugins/ipam/main.go#L43), you may need to read from `Stderr` for the error message.

File: agent/engine/docker_task_engine.go [hunk 2]
```diff
@@ -936,44 +956,59 @@ func (engine *DockerTaskEngine) isParallelPullCompatible() bool {
...
-// GetAdditionalAttributes returns back attributes of the instance when register
-// the instance into cluster
-func (engine *DockerTaskEngine) GetAdditionalAttributes() []*ecs.Attribute {
-	var attributes []*ecs.Attribute
+// getTaskENIAttribute returns the ECS "task-eni" attribute if all the necessary
+// CNI plugins are present and if they contain the capabilities required to
+// configure an ENI for a task
+func (engine *DockerTaskEngine) getTaskENIAttribute() *ecs.Attribute {
+	for _, plugin := range awsVPCCNIPlugins {
+		// Check if we can get version information from each plugin
```
Comment (richardpen) [inline agent/engine/docker_task_engine.go]: This comment seems not accurate, this part is for capability information.

File: agent/engine/docker_task_engine.go [hunk 3]
```diff
@@ -900,15 +907,28 @@ func (engine *DockerTaskEngine) Capabilities() []string {
...
 		} else {
 			seelog.Warn("Task IAM Role for Host Network not enabled due to unsuppported Docker version")
 		}
 	}
 
+	if engine.cfg.TaskENIEnabled {
+		if taskENIAttribute := engine.getTaskENIAttribute(); taskENIAttribute != nil {
+			capabilities = append(capabilities, taskENIAttribute)
+		}
+		if taskENIVersionAttribute := engine.getTaskENIPluginVersionAttribute(); taskENIVersionAttribute != nil {
```
Comment (richardpen) [inline agent/engine/docker_task_engine.go]: Do we still want to register the version if the returned `taskENIAttribute` is nil ?

File: agent/engine/docker_task_engine_test.go [hunk 4]
```diff
@@ -1195,28 +1165,84 @@ func TestCapabilities(t *testing.T) {
...
+		cniClient: cniClient,
+		cfg: &config.Config{
+			TaskENIEnabled: true,
+		},
+	}
+	attribute := taskEngine.getTaskENIAttribute()
+	assert.Nil(t, attribute)
+}
+
+func TestGetTaskENIAttributeErrorGettingCapabilitiesFromPlugin(t *testing.T) {
```
Comment (richardpen) [inline agent/engine/docker_task_engine_test.go]: Looks like you forget to edit this test, it's the same as above.

File: agent/ecscni/plugin.go [hunk 5]
```diff
@@ -187,13 +188,54 @@ func (client *cniClient) Version(name string) (string, error) {
...
+	file := filepath.Join(client.pluginsPath, name)
+
+	// Check if the plugin file exists before executing it
+	_, err := os.Stat(file)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := exec.Command(file, capabilitiesCommand)
+	capabilitiesInfo, err := cmd.Output()
```
Comment (richardpen) [inline agent/ecscni/plugin.go]: You may need to read from `Stderr` for the error message based on the implementation [here](https://github.com/aws/amazon-ecs-cni-plugins/blob/dev/plugins/ipam/main.go#L43).  Same to line 187.

File: agent/ecscni/plugin.go [hunk 6]
```diff
@@ -187,13 +188,54 @@ func (client *cniClient) Version(name string) (string, error) {
...
+
+// cniPluginVersion is used to convert the JSON output of the
+// '--version' command into a string
+type cniPluginVersion struct {
+	Version string `json:"version"`
+	Dirty   bool   `json:"dirty"`
+	Hash    string `json:"gitShortHash"`
+}
+
+func (version *cniPluginVersion) str() string {
```
Comment (richardpen) [inline agent/ecscni/plugin.go]: Can you add a comment here of how is the string looks like?

File: agent/api/ecsclient/client.go [hunk 7]
```diff
@@ -267,12 +281,52 @@ func getCpuAndMemory() (int64, int64) {
...
+
+	subnetID, err := client.ec2metadata.SubnetID(mac)
+	if err != nil {
+		seelog.Warnf("Unable to get the Subnet ID of the primary network interface: %v", err)
+		return nil, err
+	}
+
+	return []*ecs.Attribute{
+		{
+			Name:  aws.String("ecs.vpc-id"),
```
Comment (aaithal) [inline agent/api/ecsclient/client.go]: 👍

File: agent/engine/docker_task_engine.go (hunk 1 above)
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: `cmd.Output` should be capturing all of that anyway IIUC https://golang.org/src/os/exec/exec.go?s=12344:12382#L452

File: agent/engine/docker_task_engine.go (hunk 2 above)
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: i'll edit this

File: agent/engine/docker_task_engine.go (hunk 3 above)
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: Probably not. I'll refactor this a bit

File: agent/ecscni/plugin.go (hunk 6 above)
Comment (aaithal) [inline agent/ecscni/plugin.go]: 👍

File: agent/engine/docker_task_engine_test.go (hunk 4 above)
Comment (aaithal) [inline agent/engine/docker_task_engine_test.go]: Yeah, this was supposed to test the error being returned. Will fix this.

**Comment by richardpen** [comment]
Just noticed that the pause image name needs to be changed to the one we build, basically it's two lines change [task.go](https://github.com/aws/amazon-ecs-agent/blob/enis/agent/api/task.go#L47) and [docker_task_engine_test.go](https://github.com/aws/amazon-ecs-agent/blob/enis/agent/engine/docker_task_engine_test.go#L1422), could you just include that in this pr?

File: agent/app/agent_unix_test.go [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// +build !windows
+// +build linux
```
Comment (richardpen) [inline agent/app/agent_unix_test.go]: Why this was changed?

Comment (aaithal) [inline agent/app/agent_unix_test.go]: Makes it stronger. Just fortifying the intent that it will cause failures if not running on linux.

**Pull request merged**